cpu/esp32/esp-ieee802154/esp_ieee802154_hal.c

```diff
@@ -0,0 +1,461 @@
+/*
+ * Copyright (C) 2025 Gunar Schorcht
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+#include "esp_ieee802154.h"
+
+#include "esp_ieee802154_hal.h"
+#include "log.h"
+#include "net/ieee802154/radio.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+/* Although the device driver supports IEEE802154_CAP_IRQ_TX_START,
+ * IEEE802154_CAP_IRQ_RX_START and IEEE802154_CAP_IRQ_CCA_DONE, we are not
+ * using it for now to avoid unnecessary performance degradation as it is
+ * not used by any link layer driver */
+#define _USE_RX_START   0
+#define _USE_TX_START   0
+#define _USE_CCA_DONE   0
+
+#define _USE_SET_RX     1
+#define _USE_SET_IDLE   1
+
+static const ieee802154_radio_ops_t esp_ieee802154_driver;
+static ieee802154_dev_t *esp_ieee802154_dev;
+
+static uint8_t  _tx_frame[IEEE802154_FRAME_LEN_MAX + 1]; /* additional byte 0 is used for length */
+static uint8_t *_rx_frame;
+static const uint8_t *_ack_frame;
+
+static esp_ieee802154_frame_info_t *_rx_frame_info;
+static esp_ieee802154_frame_info_t *_ack_frame_info;
+
+static bool _channel_clear;
+
+static esp_ieee802154_tx_error_t _tx_error;     /* error on last transmit */
+
+static int _write(ieee802154_dev_t *dev, const iolist_t *psdu)
+{
+    (void)dev;
+
+    DEBUG("[esp_ieee802154] %s\n", __func__);
+
+    assert(psdu);
+
+    memset(_tx_frame, 0, sizeof(_tx_frame));
+
+    /* copy packet data into the _tx_frame buffer */
+    unsigned len = 0;
+    for (; psdu; psdu = psdu->iol_next) {
+        if (psdu->iol_len) {
+            memcpy(&_tx_frame[len + 1], psdu->iol_base, psdu->iol_len);
+            len += psdu->iol_len;
+        }
+    }
+
+    /* length of the package. */
+    _tx_frame[0] = len + IEEE802154_FCS_LEN;
+
+    DEBUG("[esp_ieee802154] %s: put %d bytes to _tx_frame\n", __func__, _tx_frame[0]);
+
+    return 0;
+}
+
+static int _len(ieee802154_dev_t *dev)
+{
+    (void)dev;
+    DEBUG("[esp_ieee802154] %s\n", __func__);
+    return _rx_frame[0] - IEEE802154_FCS_LEN;
+}
+
+static int _read(ieee802154_dev_t *dev, void *buf, size_t size, ieee802154_rx_info_t *info)
+{
+    (void)dev;
+
+    DEBUG("[esp_ieee802154] %s: buf=%p size=%u info=%p\n",
+          __func__, buf, size, info);
+
+    unsigned len = _rx_frame[0] - IEEE802154_FCS_LEN;
+    int res = 0;
+
+    if (size < len) {
+        DEBUG("[esp_ieee802154] %s: buffer to small (%u < %u)\n",
+              __func__, size, len);
+        res = -ENOBUFS;
+    }
+    else if (buf) {
+        DEBUG("[esp_ieee802154] %s: read packet of length %d\n", __func__, len);
+        if (info) {
+            info->rssi = ieee802154_dbm_to_rssi(_rx_frame[len]);
+            info->lqi = _rx_frame[len+1];
+        }
+        memcpy(buf, &_rx_frame[1], len);
```

Comment:
```suggestion
        memcpy(buf, _rx_frame+1, len);
```

IMO this would be a nicer syntax, I had to think quite a bit about what this does and essentially we just want to start the `memcpy` from the second byte.

You can dismiss the comment if it's tooooo nitpicky 😅